src/CustomCollection.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Letsgoi\CustomCollection;
+
+use ArrayAccess;
+use ArrayIterator;
+use Countable;
+use IteratorAggregate;
+use Letsgoi\CustomCollection\Exceptions\CustomCollectionKeyNotExistException;
+use Letsgoi\CustomCollection\Exceptions\CustomCollectionTypeErrorException;
+use Traversable;
+
+abstract class CustomCollection implements IteratorAggregate, ArrayAccess, Countable
+{
+    /** @var array */
+    protected $items;
+
+    abstract protected function getCollectionType(): string;
+
+    public function __construct(array $items = [])
+    {
+        $this->items = $items;
+
+        $this->checkItems();
+    }
+
+    /**
+     * Iterator method
+     *
+     * @return Traversable
+     */
+    public function getIterator(): Traversable
+    {
+        return new ArrayIterator($this->items);
+    }
+
+    /**
+     * Add item to collection as array
+     *
+     * @param mixed $key
+     * @param mixed $item
+     * @return void
+     * @throws CustomCollectionTypeErrorException
+     */
+    public function offsetSet($key, $item): void
+    {
+        $this->checkItemType($item);
+
+        if ($key === null) {
+            $this->items[] = $item;
+        } else {
+            $this->items[$key] = $item;
+        }
+    }
+
+    /**
+     * Get item from collection with key
+     *
+     * @param mixed $key
+     * @return mixed
+     */
+    public function offsetGet($key)
+    {
+        return $this->items[$key];
+    }
+
+    /**
+     * Checks if the collection as an item by key
+     *
+     * @param mixed $key
+     * @return bool
+     */
+    public function offsetExists($key): bool
+    {
+        return array_key_exists($key, $this->items);
+    }
+
+    /**
+     * Remove item from collection by key
+     *
+     * @param mixed $key
+     * @return void
+     */
+    public function offsetUnset($key): void
+    {
+        unset($this->items[$key]);
+    }
+
+    public function count(): int
```

Comment:
should you add the docblock? We have it in all class methods

Reply:
you're right! forgot it!